Forbid function expressions in statement context: Implementation

kazcw opened the pull request:
### Pull Request Description

Implement a syntax error when a function-creating expression (e.g. an operator section, template function, or lambda) is used as a statement--doing so is never useful, and likely to be a mistake.

This PR adds support for detecting the case to the parser; the error is not enabled yet, as #13955 must be resolved first.

### Important Notes

- This is a breaking language change: Certain types of dead code, that were previously ignored, are no longer allowed.
  - Note that this includes `//`-style "comments", which are no longer allowed in Enso code. They used to sometimes "work", but now will always trigger the suspicious-dead-code error.
- Also: Ported syntax error tests to `insta`, and extended them to cover error messages and locations.
- All syntax errors now are now defined as `SyntaxError` values. This improves maintainability, and enables more efficient parser FFI.
- More consistent syntax error punctuation.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [ ] Unit tests have been written where possible.
- [ ] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

kazcw committed:
Forbid function expressions in statement context

kazcw committed:
improve script portability and performance

JaroslavTulach commented on test/Table_Tests/src/In_Memory/Table_Spec.enso:
```diff
@@ -161,9 +161,6 @@ add_specs suite_builder =
             r4.pretty . should_equal "Table.new [['foo', [1, 2, 3]], ['bar', [False, True, False]], ['date', [Date.new 2022 8 27, Date.new 1999 1 1, Date.new 2012 1 23]], ['time', [Time_Of_Day.new 18, Time_Of_Day.new 1 2 34, Time_Of_Day.new 12]]]"
             Debug.eval r4.pretty . should_equal r4
 
-        group_builder.specify "allows setting the column type" <|
```
Detecting (and not allowing) this kind of dead code seems quite OK!

JaroslavTulach commented on lib/rust/parser/debug/tests/parse.rs:
```diff
@@ -1526,6 +1526,15 @@ fn invalid_unspaced_operator_sequence() {
+    expect_invalid_node("main =\n    \\x -> x\n    x");
+    expect_invalid_node("main =\n    _ x\n    x");
+    // Catch a common error; See: https://github.com/enso-org/enso/issues/11203
+    expect_invalid_node("main =\n    x +\n        1 +\n        2");
```
Disallowing these kinds of expressions seems OK to me.

JaroslavTulach reviewed (approved):
- as far as I can tell the _incompatible changes_ to the syntax
- demonstrated by the examples are OK
- such examples were invalid to begin with
- detecting them soon is going to increase developer productivity
- I think such an _incompatible changes_ are OK at current state of the project

kazcw committed:
Java-style comments no longer legal

kazcw committed:
Syntax error tests include details

- Syntax error tests cover parsed AST up to the `Invalid` node, and message of error.
- Updated all string-based errors to `SyntaxError`.
- Fix consistency of error message trailing punctation.

kazcw committed:
clippy

kazcw committed:
Merge branch 'develop' into wip/kw/forbid-function-expression-statements

kazcw commented on test/Base_Tests/src/Semantic/Conversion_Spec.enso:
```diff
@@ -163,7 +163,7 @@ add_specs suite_builder =
 
         group_builder.specify "should be callable with by-name arguments" <|
-            .from self=Foo that=4 first_param=2 . should_equal (Foo.Value [4, 2, 0, 0])
+            (.from) self=Foo that=4 first_param=2 . should_equal (Foo.Value [4, 2, 0, 0])
```
The original version here is technically incorrect (according to the draft spec / current AST syntax), but currently works as intended (according to the current backend semantics / old parser syntax). This PR adds a suspicious-dead-code error here, since the new diagnostics are of course based on the parser's understanding of the syntax.

For the moment I suggest we add these conservative parens to pass the new diagnostics. Once a decision has been made wrt. #13955 this type of expression may be safe without the parens in the future (or the parens may remain mandatory).

JaroslavTulach replied:
Instead of 

>  (according to the draft spec / current AST syntax

may I ask to get a link to some paragraph in https://github.com/enso-org/enso/tree/develop/docs/syntax where this is explained? Without incrementally documenting the current syntax, we never let #11462 happen.

PS: I know that most of the docs there is wrong... just find some place that is a bit suitable ... insert the description of the newest syntax there... possibly mark outdated docs with [!WARNING] as [described](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts)

kazcw replied:
You raise a good point:
- The need for these parens is not explainable by the *current syntax*
- We don't want to expose users to the *current AST syntax*, that's an implementation detail

Therefore there's really no way to introduce this error until we have decided #13955. I don't want to leave all these changes in limbo though. I'll merge this with the new error turned off; that should unblock #13986. Then once #13955 is resolved turning on the error will be a trivial change.

kazcw committed:
Typo

kazcw commented on engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/LazyAtomFieldTest.java:
```diff
@@ -97,9 +97,9 @@ public void testInfiniteListGenerator() throws Exception {
 
     both n =
         g = Lazy.generator 1
-        // IO.println "Generator is computed"
```
C-style line comments used to more-or-less work *under certain circumstances*, as a consequence of the operator section rules and ignored dead code. They now trigger the dead code error.

farmaazon commented on CHANGELOG.md:
```diff
@@ -30,13 +30,18 @@
 - [Open type check `Type&Any` lets all visible types thru][13225]
 - [Removal of --no-global-cache option][13909]
+- [Autocompletion for table expression builtin functions in table
+  expressions][13914]
```
This PR should go to frontend section. Probably messed up in #13992

JaroslavTulach replied:
I already fixed it in one of my PRs as well. Not sure whether it already merged. Probably not.

farmaazon commented on lib/rust/parser/debug/src/test.rs:
```diff
@@ -51,72 +51,75 @@ pub fn block<T: AsRef<str>>(code: T) -> String {
             }
             _ => (),
         });
         errors.into_inner()
```
Interesting, that we need Cells here. Couldn't it be a just mutable value, with mut reference captured in closure?

kazcw committed:
Review

kazcw committed:
Review

kazcw committed:
Update tests

JaroslavTulach commented on CHANGELOG.md:
```diff
@@ -30,13 +33,15 @@
 - [Register and `lookup_services` in package.yaml][11868]
 - [Open type check `Type&Any` lets all visible types thru][13225]
 - [Removal of --no-global-cache option][13909]
+- [A function expression is no longer allowed in statement context, where it
```
Shorter text like the current title _"Forbid function expressions in statement context"_ or something Enso developer possitive like _"Report syntax errors when unused code detected"_  might be a better summary of the change.

kazcw committed:
Review

kazcw committed:
Fix merge

kazcw committed:
Disable new error for now

kazcw committed:
whitespace

kazcw committed:
Merge remote-tracking branch 'origin/develop' into wip/kw/forbid-function-expression-statements
